components/brave_news/browser/resources/Feed.tsx

```diff
@@ -121,33 +125,77 @@ export default function Component({ feed }: Props) {
 
     loadMoreObserver.current.observe(el);
   }, [])
+
+  // Create intersection observer & relevant state to measure
+  // the amount of viewed cards in the session.
+  const lastViewedCardCount = React.useRef(0);
+  const viewDepthIntersectionObserver = React.useRef(new IntersectionObserver(entries => {
+    const inViewCounts = entries
+      .filter(e => e.intersectionRatio === 1)
+      .map(e => Number(e.target.getAttribute(CARD_COUNT_ATTRIBUTE)));
+    if (!inViewCounts.length) {
+      return;
+    }
+    const largestCardCount = Math.max(...inViewCounts);
+    // Ensure we only report increases in scroll depth
+    // by comparing to the last scroll card count
+    if (lastViewedCardCount.current >= largestCardCount) {
+      return;
+    }
+    const newViews = largestCardCount - lastViewedCardCount.current;
+    if (!!onSessionStart && lastViewedCardCount.current === 0 && newViews > 0) {
+      onSessionStart();
+    }
+    lastViewedCardCount.current = largestCardCount;
+    if (onViewCountChange) {
+      onViewCountChange(newViews);
+    }
+  }));
+
+  const registerViewDepthObservation = (element: HTMLElement | null) => {
+    if (!element) {
+      return;
+    }
+    viewDepthIntersectionObserver.current.observe(element);
+  };
+
   const cards = React.useMemo(() => {
     const count = Math.min(feed?.items.length ?? 0, cardCount)
+    let currentCardCount = 0;
     return feed?.items.slice(0, count).map((item, index) => {
       let el: React.ReactNode
 
       if (item.advert) {
         el = <Advert info={item.advert} />
       }
       else if (item.article) {
-        el = <Article info={item.article} />
+        el = <Article info={item.article} feedDepth={currentCardCount} />
       }
       else if (item.cluster) {
-        el = <Cluster info={item.cluster} />
+        el = <Cluster info={item.cluster} feedDepth={currentCardCount} />
       }
       else if (item.discover) {
         el = <Discover info={item.discover} />
       }
       else if (item.hero) {
-        el = <HeroArticle info={item.hero} />
+        el = <HeroArticle info={item.hero} feedDepth={currentCardCount} />
       } else {
         throw new Error("Invalid item!" + JSON.stringify(item))
       }
 
+      if (item.cluster) {
+        currentCardCount += item.cluster.articles.length;
+      } else if (!item.advert) {
+        currentCardCount++;
+      }
+
       const key = getKey(item, index)
-      return <div className={CARD_CLASS} onClickCapture={saveScrollPos(key)} key={key} data-id={key} ref={index === count - 1 ? setLastCardRef : undefined}>
-        {el}
-      </div>
+      return <>
```

Comment:
I don't think we need this fragment

Reply:
it saves us from having an outer div for each card